### PR TITLE
Option to disallow all expressions of type Any(--disallow-any=expr)

### PR DESCRIPTION
**Pull request by ilinum**
The option disallows all expressions of type Any except:
* if a value of type Any used as a second parameter to `cast`
* if a value of type Any is assigned to a variable with an explicit type annotation

### CODE REVIEW

File: test-data/unit/check-flags.test [hunk 1]
```diff
@@ -509,3 +509,50 @@ x, y = 1, 2  # type: Unchecked, Unchecked
...
+k = Foo.doo()
+
+z = cast(int, Foo().g)
+m = cast(Any, Foo().g)
+y = Foo().g
+print(x[0])
+print(x[foo(0)])
+[builtins fixtures/list.pyi]
+[out]
+main:8: error: Expressions of type "Any" are disallowed (has type "Any")
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: It'd be better to use the inline error syntax here (which you may not have seen).  Basically you can put `# E: ERROR MESSAGE` at the end of a line in a test, and it will be understood as indicating the appropriate error output on that line.

File: test-data/unit/check-flags.test [hunk 2]
```diff
@@ -509,3 +509,50 @@ x, y = 1, 2  # type: Unchecked, Unchecked
...
+main:17: error: Expressions of type "Any" are disallowed (has type "Any")
+main:20: error: Expressions of type "Any" are disallowed (has type "Any")
+main:21: error: Expressions of type "Any" are disallowed (has type "Any")
+main:23: error: Expressions of type "Any" are disallowed (has type "Any")
+
+[case testDisallowAnyExprGeneric]
+# flags: --disallow-any=expr
+from typing import List
+
+blargle: List = []
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: Nit: as much as I like `blargle`, `l` is probably better here.

File: test-data/unit/check-flags.test (hunk 1 above)
Comment (ddfisher) [inline test-data/unit/check-flags.test]: You'll have to break this up into multiple tests so you can do this (because there are multiple errors on some lines), but it'd be worthwhile.

File: mypy/checkexpr.py [hunk 3]
```diff
@@ -2429,6 +2438,19 @@ def narrow_type_from_binder(self, expr: Expression, known_type: Type) -> Type:
         return known_type
 
 
+def has_any_type(t: Type) -> bool:
+    """Whether t contains an Any type"""
+    return t.accept(HasAnyType())
+
+
+class HasAnyType(TypeQuery[bool]):
```
Comment (ddfisher) [inline mypy/checkexpr.py]: Elsewhere in the file `TypeQuery` is referred to as the partially qualified `types.TypeQuery`.  I'd do that here for consistency.

File: mypy/checkexpr.py [hunk 4]
```diff
@@ -2191,7 +2192,8 @@ def visit_backquote_expr(self, e: BackquoteExpr) -> Type:
     def accept(self,
                node: Expression,
                type_context: Type = None,
-               allow_none_return: bool = False
+               allow_none_return: bool = False,
+               disallow_any: bool = True,
```
Comment (ddfisher) [inline mypy/checkexpr.py]: I'd consider naming this `always_allow_any` (or similar) instead.  It's a bit misleading as-is.

File: mypy/checkexpr.py [hunk 5]
```diff
@@ -2211,6 +2213,13 @@ def accept(self,
         self.type_context.pop()
         assert typ is not None
         self.chk.store_type(node, typ)
+
+        if (disallow_any and
+                not self.chk.is_stub and
+                has_any_type(typ) and
+                'expr' in self.chk.options.disallow_any):
```
Comment (ddfisher) [inline mypy/checkexpr.py]: Nit: The last conditional should be first here.  It's standard (in this codebase at least) for conditionals involving flags for the flag to be checked first.  (And I think that makes the most conceptual sense, though I'm not sure I can well explain why.)

File: mypy/checkexpr.py [hunk 6]
```diff
@@ -204,7 +204,7 @@ def visit_call_expr(self, e: CallExpr, allow_none_return: bool = False) -> Type:
                         or isinstance(typ, NameExpr) and node and node.kind == nodes.TYPE_ALIAS):
                     self.msg.type_arguments_not_allowed(e)
         self.try_infer_partial_type(e)
-        callee_type = self.accept(e.callee)
+        callee_type = self.accept(e.callee, disallow_any=False)
```
Comment (ddfisher) [inline mypy/checkexpr.py]: Why are `Any` exprs allowed here?

File: mypy/messages.py [hunk 7]
```diff
@@ -6,7 +6,7 @@
 import re
 import difflib
 
-from typing import cast, List, Dict, Any, Sequence, Iterable, Tuple
+from typing import cast, List, Dict, Any, Sequence, Iterable, Tuple, Optional
```
Comment (ddfisher) [inline mypy/messages.py]: Unused import.

File: test-data/unit/check-flags.test (hunk 1 above)
Comment (ddfisher) [inline test-data/unit/check-flags.test]: Thinking some more about this: I'd like to see this split up into multiple tests anyway, each testing a specific aspect (both positive and negative) of this PR (and named appropriately).  E.g. `testDisallowAnyAllowsAnyInCast` or similar.

File: mypy/checker.py [hunk 8]
```diff
@@ -1736,7 +1736,7 @@ def check_simple_assignment(self, lvalue_type: Type, rvalue: Expression,
             # '...' is always a valid initializer in a stub.
             return AnyType()
         else:
-            rvalue_type = self.expr_checker.accept(rvalue, lvalue_type)
+            rvalue_type = self.expr_checker.accept(rvalue, lvalue_type, disallow_any=False)
```
Comment (ilinum) [inline mypy/checker.py]: Yeah, that's a good point. I will change that to `disallow_any=lvalue_type is None`. It seems like `lvalue_type` is the type from an explicit type annotation.

---

As for not allowing `Any` expressions when assigning to a variable with an explicit type annotation, here are my thoughts.

People are going to be calling untyped functions from their code. `--disallow-any=expr` needs to allow for this, otherwise no one will use it.
In my mind, there are two good ways to do it: explicit cast (call to function `cast`) and implicit cast (assignment to a variable with an explicit type annotation). 
The advantage of an explicit cast is that you can't do it accidentally.
However, the explicit cast syntax makes the code less readable.
Compare with an implicit cast:

```py
# explicit cast
result = cast(float, some_long_method_name(that, (has, many), complicated=params))
# implicit cast
result = some_long_method_name(that, (has, many), complicated=params)  # type: float
```

A big disadvantage of an implicit cast is that someone can use it accidentally. They won't look at the definition of the function they are calling and just put a type there thinking that mypy will complain if the type is wrong.

This PR has both implicit and explicit cast implemented because I think it's worth having both.
But I would be interested in hearing your reasoning for this!

File: mypy/messages.py (hunk 7 above)
Comment (ilinum) [inline mypy/messages.py]: Good catch!
I changed my PyCharm settings to complain if there's an unused import when I'm committing.

File: mypy/messages.py [hunk 9]
```diff
@@ -897,6 +897,10 @@ def typeddict_item_name_not_found(self,
     def type_arguments_not_allowed(self, context: Context) -> None:
         self.fail('Parameterized generics cannot be used with class or instance checks', context)
 
+    def disallowed_any_type(self, typ: Type, context: Context) -> None:
+        self.fail('Expressions of type "Any" are disallowed '
```
Comment (ilinum) [inline mypy/messages.py]: I'm not sure what you mean. 
Do you mean a situation when a type contains `Any` (e.g. `List[Any]`)?

File: mypy/checkexpr.py (hunk 6 above)
Comment (ilinum) [inline mypy/checkexpr.py]: This allows us to call function that are one of:
* unresolved
* unannotated or take `Any` as a parameter

Should we disallow this? My general thinking is that we want to make `--disallow-any=expr` easy to enable per-module. That way you can turn the option on for current module even if other modules are not well annotated.

As for functions that take explicitly annotated `Any` parameters, I think in a world with perfect stubs, we would disallow calling those. However, that's not the case (at least now) and, for instance, `print` function accept arguments of type `Any`. This would mean that you can't call `print` with `--disallow-any=expr` enabled.

Comment (gvanrossum) [inline mypy/checkexpr.py]: However nothing would be lost if `print`'s args were converted to `object`.

File: test-data/unit/check-flags.test (hunk 1 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: Yes, I will split them up! :)

File: test-data/unit/check-flags.test (hunk 2 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: oh well, yeah, `l` is more understandable.

File: mypy/checkexpr.py (hunk 5 above)
Comment (ilinum) [inline mypy/checkexpr.py]: Yeah, I think it's easier when you're looking through the code. 
Also, it might be a little faster.

File: mypy/checkexpr.py (hunk 3 above)
Comment (ilinum) [inline mypy/checkexpr.py]: Yep!

File: mypy/checkexpr.py (hunk 4 above)
Comment (ilinum) [inline mypy/checkexpr.py]: I was looking for a better name. `always_allow_any` is good, I think.

File: mypy/checkexpr.py (hunk 6 above)
Comment (ilinum) [inline mypy/checkexpr.py]: Did you mean updated in `typeshed`? 

Or silently treat them as `object` if this flag is enabled? I think the second is a good idea but would be really hard to implement, from a technical viewpoint.

**Comment by ilinum** [comment]
Addressed first round of code review feedback!
The following still need discussion:
* https://github.com/python/mypy/pull/3519#discussion_r121233839
* https://github.com/python/mypy/pull/3519#discussion_r121241083
* https://github.com/python/mypy/pull/3519#discussion_r121233647

File: mypy/messages.py (hunk 9 above)
Comment (ddfisher) [inline mypy/messages.py]: Yep, exactly.

File: mypy/checkexpr.py (hunk 6 above)
Comment (ddfisher) [inline mypy/checkexpr.py]: I think he means updated in typeshed. Type checking-wise, from the caller's perspective, there's no difference between an an argument of type `Any` and and argument of type `object`.

Anyway, I don't think this should be allowed.  The primary purpose of this flag is _not_ to be something that's easy to enable -- it's to provide clear, strong guarantees about how thoroughly a module is type checked.  In my mind, that sets a high bar for any exceptions we'd consider making.

File: mypy/messages.py (hunk 9 above)
Comment (ilinum) [inline mypy/messages.py]: Updated the error message to say `Expressions of type that contains type "Any" are disallowed (has type {})` if type is not `Any`.

File: mypy/checker.py (hunk 8 above)
Comment (ddfisher) [inline mypy/checker.py]: I see the same pros and cons as you -- I just weigh them differently.  Mainly, I think it's important for it to be really clear where the `Any`s are.

That said, IIRC it's not particularly common to annotate the result of a function call, so maybe the type annotation serves that purpose well enough.  Let's go with that for now.

File: mypy/checker.py [hunk 10]
```diff
@@ -1292,8 +1292,9 @@ def check_assignment(self, lvalue: Lvalue, rvalue: Expression, infer_lvalue_type
                 self.check_indexed_assignment(index_lvalue, rvalue, lvalue)
 
             if inferred:
-                self.infer_variable_type(inferred, lvalue, self.expr_checker.accept(rvalue),
-                                         rvalue)
+                is_cast = isinstance(rvalue, CallExpr) and rvalue.is_cast()
+                init_type = self.expr_checker.accept(rvalue, always_allow_any=is_cast)
+                self.infer_variable_type(inferred, lvalue, init_type, rvalue)
```
Comment (ddfisher) [inline mypy/checker.py]: What's the purpose of this change?

File: mypy/messages.py [hunk 11]
```diff
@@ -898,8 +898,12 @@ def type_arguments_not_allowed(self, context: Context) -> None:
...
 
     def disallowed_any_type(self, typ: Type, context: Context) -> None:
-        self.fail('Expressions of type "Any" are disallowed '
-                  '(has type {})'.format(self.format(typ)), context)
+        if isinstance(typ, AnyType):
+            infix = 'type'
+        else:
+            infix = 'type that contains type'
+        self.fail('Expressions of {} "Any" are disallowed '
+                  '(has type {})'.format(infix, self.format(typ)), context)
```
Comment (ddfisher) [inline mypy/messages.py]: I think this may still need some wordsmithing.  "Expressions of type that contains type 'Any' are disallowed" is not very easy to parse, IMO.

Comment (ddfisher) [inline mypy/messages.py]: Also `Expressions of type "Any" are disallowed (has type "Any")` is redundant, which could be confusing.

File: test-data/unit/check-flags.test [hunk 12]
```diff
@@ -509,3 +509,72 @@ x, y = 1, 2  # type: Unchecked, Unchecked
...
 main:4: error: Type of variable becomes "Any" due to an unfollowed import
 main:6: error: A type on this line becomes "Any" due to an unfollowed import
+
+[case testDisallowAnyExprUnannotatedFunction]
+# flags: --disallow-any=expr
+def g(s):
+    return s  # E: Expressions of type "Any" are disallowed (has type "Any")
+
+g(0)
+g('hello')
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: What's the purpose of this line?  It seems like it's testing the same thing as the previous.

File: test-data/unit/check-flags.test [hunk 13]
```diff
@@ -509,3 +509,72 @@ x, y = 1, 2  # type: Unchecked, Unchecked
 [out]
 main:4: error: Type of variable becomes "Any" due to an unfollowed import
 main:6: error: A type on this line becomes "Any" due to an unfollowed import
+
+[case testDisallowAnyExprUnannotatedFunction]
+# flags: --disallow-any=expr
+def g(s):
+    return s  # E: Expressions of type "Any" are disallowed (has type "Any")
+
+g(0)
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: It'd be nice to put a short comment on this line explaining what it's testing and why it's not an error.

File: test-data/unit/check-flags.test [hunk 14]
```diff
@@ -509,3 +509,72 @@ x, y = 1, 2  # type: Unchecked, Unchecked
...
+g('hello')
+
+v = g(1)  # E: Expressions of type "Any" are disallowed (has type "Any")
+w: int = g(1)
+
+[case testDisallowAnyExprExplicitAnyParam]
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: Not sure I fully understand what this test is testing.

File: test-data/unit/check-flags.test [hunk 15]
```diff
@@ -509,3 +509,72 @@ x, y = 1, 2  # type: Unchecked, Unchecked
...
+[builtins fixtures/list.pyi]
+
+[case testDisallowAnyExprAllowsAnyInCast]
+# flags: --disallow-any=expr
+from typing import Any, cast
+class Foo:
+    g: Any = 2
+
+z = cast(int, Foo().g)
+m = cast(Any, Foo().g)
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: I don't think this should be allowed.  You won't be able to use `m` without error!

File: test-data/unit/check-flags.test [hunk 16]
```diff
@@ -509,3 +509,72 @@ x, y = 1, 2  # type: Unchecked, Unchecked
...
+[builtins fixtures/list.pyi]
+
+[case testDisallowAnyExprAllowsAnyInVariableAssignmentWithExplicitTypeAnnotation]
+# flags: --disallow-any=expr
+from typing import Any
+class Foo:
+    g: Any = 2
+
+z: int = Foo().g
+m: Any = Foo().g
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: This probably shouldn't be allowed either.  Also, could you add a comment-style annotation to this test?

File: test-data/unit/check-flags.test [hunk 17]
```diff
@@ -509,3 +509,72 @@ x, y = 1, 2  # type: Unchecked, Unchecked
...
+[builtins fixtures/list.pyi]
+
+[case testDisallowAnyExprAllowsNonExistentMembers]
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: Is this test really unique?  Given the behavior of the other tests, it seems like this is primarily testing that calling a nonexistent function returns `Any` after the error, which doesn't need to be tested here.

File: mypy/checkexpr.py (hunk 6 above)
Comment (ilinum) [inline mypy/checkexpr.py]: That means we should probably go through all stubs in typeshed and parameters of type `Any` to `object`. I can look into it. Just to confirm, it shouldn't break anything, right?

I see your point.
I still think we should not give an error if a user is calling a function that takes `Any`. Often, they can't modify the stubs of the function they are calling.
That being said, I'm okay with disallowing this and then seeing if it's a problem for the users.

How should we handle unresolved functions? (I imagine this would be the case when `--disallow-any=expr` is used alongside `--ignore-missing-imports`)

File: mypy/checker.py (hunk 10 above)
Comment (ilinum) [inline mypy/checker.py]: For the following case: 
```py
m = cast(Any, Foo().g)
```

After splitting up the test cases like you suggested, I realized that that case doesn't work.

File: mypy/messages.py (hunk 11 above)
Comment (ilinum) [inline mypy/messages.py]: How about `Expression type contains "Any" (has type {})`?
If the actual type is `Any`, we can say `Expressions has type "Any"`.

File: test-data/unit/check-flags.test (hunk 12 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: Yep, it does test the same thing. I will remove this line.

File: test-data/unit/check-flags.test (hunk 13 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: After what we discussed [here](https://github.com/python/mypy/pull/3519/files/6517ae4055199642a7d3f8b52f81e7b37458995e#r121241083), it will be an error.

File: test-data/unit/check-flags.test (hunk 15 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: Good point, will change.

File: test-data/unit/check-flags.test (hunk 16 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: Good point, will change.

File: mypy/checker.py (hunk 10 above)
Comment (ilinum) [inline mypy/checker.py]: Actually, I am going to roll back this particular change because of what we discussed [here](https://github.com/python/mypy/pull/3519#discussion_r121779999).

File: test-data/unit/check-flags.test (hunk 14 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: I think it's the same as `testDisallowAnyExprUnannotatedFunction` except this time function is annotated but type parameter is explicitly `Any`.

File: test-data/unit/check-flags.test (hunk 17 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: Yes, you are right. I will remove the test :)

**Comment by ilinum** [comment]
I pushed some commits that should address almost all of second round of code review.
What we have left on this PR:
[ ] Decide on error message [wording](https://github.com/python/mypy/pull/3519#discussion_r121778948).
[ ] How to handle unresolved references (https://github.com/python/mypy/pull/3519#discussion_r121787747).

File: test-data/unit/check-flags.test [hunk 18]
```diff
@@ -509,3 +509,69 @@ x, y = 1, 2  # type: Unchecked, Unchecked
...
+    pass
+
+def g(s: List[Any]) -> None:
+    pass
+
+f(0)
+
+# because expected type is List[Any] the inferred type of [''] is List[Any], which causes
+# more than one error on the line below
+g([''])
```
Comment (ilinum) [inline test-data/unit/check-flags.test]: As stated in the comment above, expecged type is `List[Any]`, which causes the inferred type of `['']` to be `List[Any]`. 

Honestly, this behavior seems correct but user gets a whole bunch of errors that are hard to understand. 

@ddfisher what do you think?

File: docs/source/command_line.rst [hunk 19]
```diff
@@ -290,6 +290,13 @@ Here are some more useful flags:
   of the parameters or the return type). ``unannotated`` option is
   interchangeable with ``--disallow-untyped-defs``.
 
+  ``expr`` disallows all expressions in the module that have type ``Any``.
+  If an expression of type ``Any`` appears anywhere in the module
+  mypy will output an error unless the expression is immediately
+  used as an argument to ``cast`` or assigned to a variable with an
+  explicit type annotation. Note that declaring a variable of type ``Any``
+  or casting to type ``Any`` is not allowed.
```
Comment (graingert) [inline docs/source/command_line.rst]: should this be "is allowed"

Comment (ilinum) [inline docs/source/command_line.rst]: No, `cast(Any, foo)` and `x: Any = foo` are not allowed (we discussed it [here](https://github.com/python/mypy/pull/3519#discussion_r121779999) and [here](https://github.com/python/mypy/pull/3519#discussion_r121780369)).

What do you think of the wording?

File: mypy/messages.py (hunk 11 above)
Comment (ddfisher) [inline mypy/messages.py]: That's great!

File: docs/source/command_line.rst (hunk 19 above)
Comment (ddfisher) [inline docs/source/command_line.rst]: LGTM.

File: mypy/checkexpr.py [hunk 20]
```diff
@@ -993,6 +993,10 @@ def check_arg(self, caller_type: Type, original_caller_type: Type,
                 return
             messages.incompatible_argument(n, m, callee, original_caller_type,
                                            caller_kind, context)
+        elif ('expr' in self.chk.options.disallow_any and
+                has_any_type(callee_type) and
+                not self.chk.is_stub):
+            messages.call_any_parameter(n, callee.name, callee_type, context)
```
Comment (ddfisher) [inline mypy/checkexpr.py]: Why does this need to be special cased?  Shouldn't these be caught from other places?

File: mypy/checkexpr.py (hunk 6 above)
Comment (ddfisher) [inline mypy/checkexpr.py]: It wouldn't break anything, but that's not a change we want to make: a function taking `Any` and a function taking `object` implies pretty different things about what that function supports.

I think it's reasonable for now to disallow functions that have `Any` as an argument.  We may need to relax this requirement later, but it's easier to start strict and then remove errors if they're too much.

Unresolved functions should definitely be errors in this mode.  They're unambiguously expressions with type `Any`.

File: mypy/checkexpr.py [hunk 21]
```diff
@@ -185,7 +185,7 @@ def visit_call_expr(self, e: CallExpr, allow_none_return: bool = False) -> Type:
         """Type check a call expression."""
         if e.analyzed:
             # It's really a special form that only looks like a call.
-            return self.accept(e.analyzed, self.type_context[-1])
+            return self.accept(e.analyzed, self.type_context[-1], always_allow_any=True)
```
Comment (ddfisher) [inline mypy/checkexpr.py]: What, exactly, is this allowing?  At a minimum, I think this needs a comment about why `always_allow_any=True`.

File: mypy/checkexpr.py [hunk 22]
```diff
@@ -204,7 +204,7 @@ def visit_call_expr(self, e: CallExpr, allow_none_return: bool = False) -> Type:
                         or isinstance(typ, NameExpr) and node and node.kind == nodes.TYPE_ALIAS):
                     self.msg.type_arguments_not_allowed(e)
         self.try_infer_partial_type(e)
-        callee_type = self.accept(e.callee)
+        callee_type = self.accept(e.callee, always_allow_any=True)
```
Comment (ddfisher) [inline mypy/checkexpr.py]: Is `always_allow_any=True` here so we can give a more specific error message in `check_arg`?  If so, could you add a comment about that?

File: test-data/unit/check-flags.test (hunk 18 above)
Comment (ddfisher) [inline test-data/unit/check-flags.test]: I've changed my mind on the functions-with-arguments-of-type-Any issue -- I think it's too hard for users to understand, and isn't necessarily in the sprit of this flag.  I think it'd be better to allow those instead.  That should help with a lot of confusing errors.

Comment (ilinum) [inline test-data/unit/check-flags.test]: Just a note that `g([''])` is still an going to be an error because type of `['']` is inferred with expected type of `List[Any]`. and `['']` technically has type `List[Any]`

File: mypy/checkexpr.py (hunk 21 above)
Comment (ilinum) [inline mypy/checkexpr.py]: I'm not totally sure. I removed it and none of the tests failed, so I am going to remove it. I also played around with it a little bit and didn't find any difference.

I did some investigation what it could be and I think it was me trying to allow `cast(Any, foo)`. 
But since then we decided to not allow that.

File: mypy/checkexpr.py (hunk 22 above)
Comment (ilinum) [inline mypy/checkexpr.py]: Yes it is. 
I think we are not giving any errors for calls to function that accept `Any` anymore.

Should I still add a comment?

Comment (ddfisher) [inline mypy/checkexpr.py]: I think it's still worth a quick comment, yeah.

File: test-data/unit/check-flags.test [hunk 23]
```diff
@@ -515,3 +515,59 @@ x, y = 1, 2  # type: Unchecked, Unchecked
 [out]
 main:4: error: Type of variable becomes "Any" due to an unfollowed import
 main:6: error: A type on this line becomes "Any" due to an unfollowed import
+
+[case testDisallowAnyExprUnannotatedFunction]
+# flags: --disallow-any=expr
+def g(s):
+    return s  # E: Expression has type "Any"
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: Sorry for not noticing this earlier, but I don't think we should give this warning in unannotated functions.  Should be easy to fix, though!

File: test-data/unit/check-flags.test [hunk 24]
```diff
@@ -515,3 +515,59 @@ x, y = 1, 2  # type: Unchecked, Unchecked
 [out]
 main:4: error: Type of variable becomes "Any" due to an unfollowed import
 main:6: error: A type on this line becomes "Any" due to an unfollowed import
+
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: It'd be nice to have some tests for the basic functionality.  These tests do a good job of exercising the more interesting edge cases, but it's worth testing putting an `Any` type in a variety of expressions to make sure it's caught/the behavior is what we expect.

File: test-data/unit/check-flags.test (hunk 23 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: Yes, not reporting errors in unannotated functions makes sense.
Do you think it's worth mentioning it in documentation?

File: test-data/unit/check-flags.test (hunk 24 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: [Done!](https://github.com/python/mypy/pull/3519/commits/52b9206294769d65c0b3b524c0d5429c23911ba1)

Let me know if you have any other test ideas.

File: mypy/checkexpr.py (hunk 22 above)
Comment (ilinum) [inline mypy/checkexpr.py]: After we talked about this in person, we decided that adding info about calling untyped functions to documentation would be enough.

**Comment by ilinum** [comment]
Thanks for reviewing so many changed and iterations!
Addressed the most recent round of code review.

**Comment by ddfisher** [comment]
Looks good!  Thanks again for bearing through the long review! 🎉

File: test-data/unit/check-flags.test (hunk 23 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: @ddfisher I think Github may have hidden this comment. I think that's something worth including in documentation. 

What's your opinion on this?
